design/k8s_secrets_labels_design.md

```diff
@@ -0,0 +1,218 @@
+
+### Design
+
+|      | Fetch/Update K8s Secret                                      | Pros                                                         | Cons                                                         | UX                                                           | Effort estimation |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ----------------- |
+| 1    | Get K8s secrets with the names specified in `K8S_SECRETS`    | - No code changes  - Backwards compatible                    | - Repetitive work <br />- Not scalable                       | List K8S Secrets one-by-one                                  | Free              |
+| 2    | Read all K8s Secrets that are marked with labels specified in `K8S_SECRETS_LABEL` env var | - Scalable grouping  - Less intervention on our part as a customer can define which types of secrets we should parse  - K8s native, as-is solution | - Not general solution so we are still K8s bound - Demands SA has "list" privileges on secrets | Add "list" to SA `verbs: ["get", "update", "list"]` <br />- List label grouping in Secrets Provider manifest<br />- Label key `conjur-secrets-label` will need to be added to each K8s Secret | 5 days            |
+
+*Decision*: Solution #2, Read all K8s Secrets that are marked with labels. It would be the most K8s native and avoids the repetitive work of having to list all the K8s Secrets each app needs. 
+
+### Customer Experience 
+
+A customer would add a label on a K8s Secret like so:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-secret
+  labels:
+    conjur-secrets-label: prod # label name can be configurable
+  type: Opaque
+stringData:
+  conjur-map: |-
+    username: secrets-accessors/db_username
+    password: secrets-accessors/db_password
+```
+
+Secrets Provider Manifest with the `K8S_SECRETS_LABEL` environment variable:
+
+```yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: secrets-provider-job
+  namespace: app-namespace
+spec:
+  template:
+    spec:
+      serviceAccountName: secret-provider-sa
+      containers:
+      - image: secrets-provider:latest
+        name: cyberark-secrets-provider-1
+        env:
+        - name: SECRETS_DESTINATION
+          value: k8s_secrets
+
+        - name: K8S_SECRETS_LABEL
+          value: conjur-secrets-label:prod
+        
+        # K8s authn-client configurations (CONJUR_APPLIANCE_URL, CONJUR_AUTHN_URL, CONJUR_ACCOUNT, CONJUR_SSL_CERTIFICATE) 
+```
+
+These labels will have an OR inclusive relationship, not an AND relationship. The label key will not be configurable and will have to be `conjur-secrets-label`. In other words, to use this enhancement, they will need to add `conjur-secrets-label` as an entry under labels.
```

Comment:
i am confused. the label is not configurable? then what is `conjur-secrets-label:prod`? what is `prod` in this case? is there another valid value to this label?

If not, let's change it to `true`.

Also - if it's not configurable, then why does the customer need to add an env var? why can't we look for it by ourselves without the customer needing to add this value?

Reply:
or do you mean that the value of this label _is_ configurable? where a customer will have one secrets-provider with `conjur-secrets-label:prod` and one with `conjur-secrets-label:staging`?